src/main/java/org/plumelib/util/ArraySet.java

```diff
@@ -0,0 +1,414 @@
+package org.plumelib.util;
+
+import java.util.AbstractSet;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.Consumer;
+import org.checkerframework.checker.index.qual.GTENegativeOne;
+import org.checkerframework.checker.index.qual.IndexOrHigh;
+import org.checkerframework.checker.index.qual.LengthOf;
+import org.checkerframework.checker.index.qual.LessThan;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.lock.qual.GuardSatisfied;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.checker.signedness.qual.UnknownSignedness;
+import org.checkerframework.dataflow.qual.Pure;
+import org.checkerframework.dataflow.qual.SideEffectFree;
+
+/**
+ * A set backed by an array. It permits null values.
+ *
+ * <p>Compared to a HashSet or LinkedHashSet: For very small sets, this uses much less space, has
+ * comparable performance, and (like a LinkedHashSet) is deterministic, with elements returned in
+ * the order they were inserted. For large sets, this is significantly less performant than other
+ * set implementations.
+ *
+ * <p>Compared to a TreeSet: This uses somewhat less space, and it does not require defining a
+ * comparator. This isn't sorted. For large sets, this is significantly less performant than other
+ * set implementations.
+ *
+ * <p>Other ArraySet implementations include:
+ *
+ * <ul>
+ *   <li>https://docs.oracle.com/en/database/oracle/oracle-database/19/olapi/oracle/olapi/ArraySet.html
+ *   <li>https://developer.android.com/reference/android/util/ArraySet
+ *   <li>https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CopyOnWriteArraySet.html
+ * </ul>
+ *
+ * All of those use the GPL or the Apache License, version 2.0, whereas this implementation is
+ * licensed under the more libral MIT License. In addition, some of those implementations forbid
+ * nulls or nondeterministically reorder the contents, and others don't specify their behavior
+ * regarding nulls and ordering. CopyOnWriteArraySet uses an ArrayList, not an array, as its
+ * representation.
+ *
+ * @param <E> the type of the set elements
+ */
+@SuppressWarnings({
+  "index", // TODO
+  "keyfor", // https://tinyurl.com/cfissue/4558
+  "lock", // not yet annotated for the Lock Checker
+  "nullness", // temporary; nullness is tricky because of null-padded arrays
+  "signedness:argument", // unannotated JDK methods; TODO: remove after CF release 3.26.1
```

Comment:
According to the TODO comment here and below, this warning suppression key can be removed.